src/keygen.rs

```diff
@@ -53,47 +53,75 @@ impl PrimeSampable for BigInt {
     }
 
     fn sample_safe_prime(bitsize: usize) -> Self {
-        // q = 2p + 1;
         let two = BigInt::from(2);
+        let four = &two + &two;
         loop {
-            let q = PrimeSampable::sample_prime(bitsize);
-            let p = (&q - BigInt::one()).div_floor(&two);
-            if is_prime(&p) {
-                return q;
-            };
+            // q = 2p + 1;
+            // We want to ensure p,q are both prime.
+            let mut q = Self::sample(bitsize);
+
+            // We flip the LSB to make sure the candidate is odd.
+            //  BitManipulation::set_bit(&mut candidate, 0, true);
+            BigInt::set_bit(&mut q, 0, true);
+
+            // We flip the 2nd LSB to make sure the candidate is 3 mod 4 because
+            // (q - 1) / 2 must also be an odd prime.
+            BigInt::set_bit(&mut q, 1, true);
+
+            // To ensure the appropiate size
+            // we set the MSB of the candidate.
+            BitManipulation::set_bit(&mut q, bitsize - 1, true);
```

Comment:
Just for unification matters, I'd suggest to change this line to look similar to two `set_bit` calls above:
```suggestion
            BigInt::set_bit(&mut q, bitsize - 1, true);
```

Note also that there's a most native and shorter form of calling `set_bit`, but that's just a matter of choice:
```rust
q.set_bit(bitsize - 1, true);
```